[AIRFLOW-4315] Add monitoring API's to airflow

raviagarwalunravel opened the pull request:
Make sure you have checked all steps below.

### Jira

- [x] My PR addresses the following Airflow Jira issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
- https://issues.apache.org/jira/browse/AIRFLOW-4315


### Description

Following endpoint have been added using this PR:
- GET /api/experimental/dags
- GET /api/experimental/dags/<DAG_ID>
- GET /api/experimental/dags/<DAG_ID/tasks
- GET /api/experimental/dag_runs
- GET /api/experimental/task_instances
- GET /api/experimental/dags/<DAG_ID>/dag_runs/EXECUTION_DATE/tasks/<TASK_ID>/logs

Following endpoints have been modified to either improve response or to add more filters:
- GET /api/experimental/dags/<DAG_ID>/tasks/<TASK_ID>
- GET /api/experimental/dags/<DAG_ID>/dag_runs
- GET /api/experimental/dags/<DAG_ID>/dag_runs/<string:execution_date>
- GET /api/experimental/dags/<DAG_ID>/dag_runs/<EXECUTION_DATE>/tasks/<TASK_ID>

### Tests

- [x] My PR adds the following unit tests
   1. tests/www/api/experimental/test_endpoints.py 

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

raviagarwalunravel committed:
[AIRFLOW-4315] Add /dags endpoint

A root level endpoint to get a list of all dags in the system.
Also add is_paused filter to /dags endpoint. This restricts the list
of dags based on value.
Possible values for is_paused : true or false.

raviagarwalunravel committed:
[AIRFLOW-4315] Add /dags/<Dag_id>/tasks endpoint

Returns all tasks associated with the dag_id.

raviagarwalunravel committed:
[AIRFLOW-4315] Modify /dags/<Dag_id>/dag_runs endpoint

More filters have been added to the endpoint to ease searching for
dag_runs.

Add state_not_equal parameter, which will filter the list and remove
all dag_runs equal to the value.
Add execution_date_before parameter, which will filter and only list
the runs before the provided date.
Add execution_date_after parameter, will will filter and only list
the runs after the provided date.
Add end_date to responese for /dags/<dag_id>/dag_runs object

raviagarwalunravel committed:
[AIRFLOW-4315] Add endpoint /dag_runs

This is a new base level endpoint to get details for any dag_runs
and can be free of any particular DAG_ID

This endpoint also has 5 filters added to it as query parameters:
- state : find dag_runs with a particular state
- state_not_equal: find dag_runs with state not equal to the given
                   state
- execution_date_before : find dag_runs with execution_date before
                          given date
- execution_date_after : find dag_runs with execution_date after
                         given date
- dag_id : find dag_runs associated with a particular dag_id

raviagarwalunravel committed:
[AIRFLOW-4315] Modify endpoint for Task_id

Modify /dags/<Dag_id>/dag_runs/<execution_date>/tasks/<Task_id>
Improve response object
- add upstream task_instances
- add downstream task_instances
- add number of attemps -> try_number

raviagarwalunravel committed:
[AIRFLOW-4315] Add endpoint /task_instances

This is a new base level endpoint to get details for any
task_instances and can be free of any particular DAG_ID or task_id.
This endpoint also has 6 filters added to it as query parameters:
- state : find task_instance with a particular state
- state_not_equal: find task_instance with state not equal to the
                   given state
- execution_date_before : find task_instance with execution_date
                          before given date
- execution_date_after : find task_instance with execution_date
                         after given date
- dag_id : find task_instance associated with a particular dag_id
- task_id : find task_instance associated with a particular task_id

raviagarwalunravel committed:
[AIRFLOW-4315] Add tests for the new/modified endpoints

mik-laj commented on tests/www/api/experimental/test_endpoints.py:
```diff
@@ -77,6 +140,24 @@ def test_task_info(self):
+
+        self.assertIn('"upstream_task_ids"', response.data.decode('utf-8'))
+        expected_upstream_task_list = ["runme_2", "runme_1", "runme_0"]
+        self.assertItemsEqual(sorted(response_content['upstream_task_ids']),
```
This method does not exist in Python 3.6. I suspect you wanted to use `assertEqual`.

raviagarwalunravel replied:
I can see a equivalent method in python 3.6 as assertCountEqual, but wont these test cases fail for python 2.x then? or do the tests not care about earlier python versions?

ashb replied:
`assertListEqual` given you are sorting the items anyway?

raviagarwalunravel committed:
[AIRFLOW-4315] Add endpoint to query logs for tasks

The endpoint is accessible at
/dags/<Dag_id>/dag_runs/<Execution_date>/tasks/<Task_id>/logs

raviagarwalunravel committed:
[AIRFLOW-4315] Add /dags/<Dag_id> endpoint

Endpoint to get information about a specfic dag_id.

raviagarwalunravel committed:
[AIRFLOW-4315] Modify /dags/<dag_id>/dag_runs/<execution_date>

Before this commit the only value returned was the state of the
dag_run. This commit adds more information to the response object.

raviagarwalunravel committed:
[AIRFLOW-4315] Fix Flake8 errors

raviagarwalunravel committed:
[AIRFLOW-4315] Modify /dags/<Dag_id>/tasks/<Task_id>

Add upsteream and downstream task information to the endpoint for
better understanding of the task in relation to the dag as a whole.

ashb commented on airflow/api/common/experimental/get_all_dag_runs.py:
```diff
@@ -0,0 +1,59 @@
+    state_not_equal = state_not_equal.lower() if state_not_equal else None
+    execution_date_before = timezone.parse(execution_date_before) if execution_date_before else None
+    execution_date_after = timezone.parse(execution_date_after) if execution_date_after else None
+    for run in DagRun.find(dag_id=dag_id, state=state, state_not_equal=state_not_equal,
```
This has a lot of duplication with airflow/api/common/experimental/get_dag_runs.py - instead of adding a new file that should be extended with the date filters and making dag_id optional.

raviagarwalunravel replied:
done

ashb commented on airflow/api/common/experimental/get_dags.py:
```diff
@@ -0,0 +1,60 @@
+            'task_count': dag.task_count,
+            'task_ids': dag.task_ids,
+            'resources': {
+                'dag_runs': configuration.conf.get('webserver', 'BASE_URL') +
```
Use `url_for`

raviagarwalunravel replied:
done.

ashb commented on airflow/api/common/experimental/get_task.py:
```diff
@@ -38,3 +38,19 @@ def get_task(dag_id, task_id):
+            'upstream_task_ids': list(task._upstream_task_ids),
+            'downstream_task_ids': list(task._downstream_task_ids)})
+    except AirflowException as err:
+        raise err
```
```suggestion
        raise
```

will preserve the stack traces better

ashb replied:
Just remove the `try`/`except` entirely and let any error be thrown.

raviagarwalunravel replied:
done.

ashb commented on airflow/api/common/experimental/get_task_logs.py:
```diff
@@ -0,0 +1,91 @@
+                         .format(task_id, execution_date))
+        raise TaskInstanceNotFound(error_message)
+
+    ti = TaskInstance.find(dag_id=dag_id, task_id=task_id, execution_date=execution_date)[0]
```
Everything in the fn up to this point should be replaced with a call to `get_task_instance(dag_id, task_id, executition_date)` from airflow.api.common.experimental.get_task_instance

raviagarwalunravel replied:
done.

ashb commented on airflow/api/common/experimental/get_dag.py:
```diff
@@ -0,0 +1,38 @@
+            return dag
+
+    error_message = "Dag id {} not found".format(dag_id)
+    raise AirflowException(error_message)
```
```suggestion
    raise DagNotFound(error_message)
```

raviagarwalunravel replied:
done.

ashb commented on airflow/api/common/experimental/get_dags.py:
```diff
@@ -0,0 +1,60 @@
+    Returns a list of all Dags
+    :return: List of all DAGs
+    """
+    dagbag = DagBag()
```
Constructing the dag bag involves parsing all the dags - this is not what we want to do. Instead we should look at the models.DAG.

raviagarwalunravel replied:
done.

ashb commented on airflow/api/common/experimental/get_dag.py:
```diff
@@ -0,0 +1,38 @@
+    dag_list = get_dags()
+
+    for dag in dag_list:
+        print(dag['dag_id'] + ' : ' + dag_id)
```
Prints will be spammy to logs - remove.

raviagarwalunravel replied:
removed.

ashb commented on airflow/api/common/experimental/get_dag.py:
```diff
@@ -0,0 +1,38 @@
+    :param dag_id: String identifier of a DAG
+    :return: DAG info for a specific DAG ID.
+    """
+    dag_list = get_dags()
```
This is a very expensive operation to get a single known/specific DAG and will make the API endpoint slow. Please re-work this function to directly get the one dag we want from the DB.

raviagarwalunravel replied:
done

ashb commented on airflow/api/common/experimental/get_task_logs.py:
```diff
@@ -0,0 +1,91 @@
+        ti=ti,
+        try_number="all" if try_number is None else try_number,
+        filename_template=filename_template)
+    return send_file(file_obj, as_attachment=True, attachment_filename=attachment_filename)
```
This method won't cope with the file being on remote storage and no longer on the worker.

Check what we do for the get_logs_with_metadata and use that logic from both places (API and WWW)

ashb commented on airflow/api/common/experimental/get_tasks.py:
```diff
@@ -34,11 +33,5 @@ def get_dag_run_state(dag_id, execution_date):
-
-    return {'state': dagrun.get_state()}
+    # Return the task.
+    return dag.task_ids
```
Why do we need a function for `get_dag(dag_id).task_ids`? That seems... verbose.

raviagarwalunravel replied:
What do you mean? We need to return the task_ids and make sure that the dag_id passed does exist. I don't understand this comment fully.

ashb replied:
I mean that we remove this function, and where it is currently called in your PR:
```python
        task_list = list()
        task_ids = get_tasks(dag_id)
```

change it to something like.

```python
        task_ids = get_dag(dag_id).task_ids
```

raviagarwalunravel replied:
What you are saying is correct, but then we would have to get DagBag in the endpoints.py file and also do the verification of dag_id existance. I was also consiously trying not to touch models from this layer, and defer to all DB access tasks to the helper function defined in the api/common/experimental folder. This implementation seems cleaner to me, but if you guys would say otherwise I will change it.

ashb replied:
The change as I proposed won't need any changes anywhere else. Try it? I don't why you'd need to touch the DagBag at all - `get_dag()` will return a DAG object or will thrown an exception

ashb commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -41,6 +47,124 @@
+    """
+    Return the list of all dag_runs
+    :query param state: a query string parameter '?state=queued|running|success...'
+    :query param state_not_equal: a query string parameter '?state_not_equal=queued|running|success...'
```
```suggestion
    :query param state__ne: a query string parameter '?state_not_equal=queued|running|success...'
```

To follow the Django "convention" of `${field}__${op}` - this applies to many other places too please.

raviagarwalunravel replied:
done in all places.

ashb commented on airflow/models/dagrun.py:
```diff
@@ -113,8 +113,9 @@ def refresh_from_db(self, session=None):
+             execution_date_before=None, execution_date_after=None,
              state=None, external_trigger=None, no_backfills=False,
-             session=None):
+             session=None, state_not_equal=None):
```
Make session the last param always please.

raviagarwalunravel replied:
done.

ashb commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -41,6 +47,124 @@
 
 
+@api_experimental.route(
+    '/dags/<string:dag_id>/dag_runs/<string:execution_date>/tasks/<string:task_id>/logs',
```
This fn should take an (optional) try_number too.

raviagarwalunravel replied:
Should the try_number be added as a query parameter?

ashb replied:
Sure, sounds good.

ashb commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -41,6 +47,124 @@
+        response = jsonify(error="{}".format(err))
+        response.status_code = err.status_code
+        return response
+    except ValueError:
```
This error handling is overly broad and could catch a ValueError from something else.

Additionally: I would like to see the conversion from string to dt object here in this FN, not in side the common api part.

raviagarwalunravel replied:
done.

ashb commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -41,6 +47,124 @@
+@requires_authentication
+def dag_runs_filter():
+    """
+    Return the list of all dag_runs
```
```suggestion
    Return the list of all dag_runs
    
```

Need a blank line between text and params for it to be rendered properly.

raviagarwalunravel replied:
done.

ashb reviewed (changes requested):
There's a bit too much duplication of the common.api methods for my taste - please make them "call" each other a bit more.

Additionally I don't like the pattern (that already existed, so not your fault) of one function per module where the function and the module share the same name.

Instead could you re-organise to things like ` airflow/api/common/experimental/dag_runs.py` that contains a few functions to do with DagRun objects for the API etc.

Could you add an `:rtype:` do the docs to say what type of object(s) the api.common.* methods return? (`dict`, `List[TaskInstance]`, `list[dict]` etc.)

alejandrounravel commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -255,18 +413,14 @@ def dag_run_status(dag_id, execution_date):
-        response.status_code = err.status_code
+    if not dagruns:
+        error_message = "No Dag run found with provided execution date"
+        response = jsonify(error="{}".format(error_message))
```
Can simplify this with just ```jsonify(error=error_message)```?
This happens in multiple places.

raviagarwalunravel replied:
yea, this is done where ever it was possible to do so.

alejandrounravel commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -150,23 +284,39 @@ def get_dag_code(dag_id):
+
+    except AirflowException as err:
+        _log.info(err)
+        response = jsonify(error="{}".format(err))
```
Can simplify this with just ```jsonify(error=error_message)```?

raviagarwalunravel replied:
This is not possible, as these statements throw error messages stating type error, something like the following :
TypeError: Object of type 'DagNotFound' is not JSON serializable

alejandrounravel commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -150,23 +284,39 @@ def get_dag_code(dag_id):
+        response.status_code = err.status_code
+        return response
+
+    return jsonify(task_list)
```
Good practice to initialize ```info = None``` just before the try-except since the return statement depends on it, or to return the info object inside of the try-block while the final return statement can just return None.

raviagarwalunravel replied:
where is info? but I have indeed moved the `task_list = list()` statement above the try block.

alejandrounravel replied:
This was a copy/paste of the same thing, please initialize ```task_list = list()``` just before the "try"

alejandrounravel commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -41,6 +47,124 @@
+        response = jsonify(error="{}".format(err))
+        response.status_code = 404
+        return response
+    return jsonify(dag)
```
Good practice to initialize ```dag = None``` just before the try-except since the return statement depends on it, or to return the dag inside of the try-block while the final return statement can just return None.

raviagarwalunravel replied:
good point.

alejandrounravel commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -115,13 +239,23 @@ def dag_runs(dag_id):
     :query param state: a query string parameter '?state=queued|running|success...'
+    :query param state_not_equal: a query string parameter '?state_not_equal=queued|running|success...'
+    :query param execution_date_before: a query string parameter to find all runs before provided date,
+    should be in format "YYYY-mm-DDTHH:MM:SS", for example: "2016-11-16T11:34:15".'
```
Nitpick: Comments have a superfluous .' at the end

raviagarwalunravel replied:
removed.

alejandrounravel commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -41,6 +47,124 @@
+    :query param is_paused: a query string parameter '?is_paused=true|false'
+    :return: List of all DAGs
+    """
+    is_paused = request.args.get('is_paused')
```
Perhaps add a robustness check here to ensure that is_paused is either true or false (not sure if str or boolean). What if the user enters an erroneous value?

raviagarwalunravel replied:
if the user enters erroneous value, it would be assumed to be false. Basically, anything other than 'true' is false.

alejandrounravel replied:
What is the type of the variable? If it's a boolean then it's good, but if it's a string, then may need to compare it using ```val.lower() == "true"```

alejandrounravel commented on airflow/models/taskinstance.py:
```diff
@@ -215,6 +215,49 @@ def try_number(self, value):
+        TI = TaskInstance
+
+        query = session.query(TI)
+        if dag_id:
```
Shouldn't at least dag_id be required?

raviagarwalunravel replied:
I guess not, what if the user is simply interested in all the data?

ashb replied:
"What if"s should not be implemented. Have a concrete use case or don't add the code.

alejandrounravel reviewed (changes requested):
Requesting changes related to robustness and code paths that always return initialized variables.

ashb commented on airflow/api/common/experimental/get_dag.py:
```diff
@@ -0,0 +1,38 @@
+from airflow.api.common.experimental.get_dags import get_dags
+
+
+def get_dag(dag_id):
```
This function has no return so cannot be working. Is it actually called from anywhere?

ashb replied:
Looks like from the get_dag_info endpoint. Given I don't think this is working you need to improve your test coverage.

ashb replied:
Oh no I see. Not at the end is all.

raviagarwalunravel committed:
[AIRFLOW-4315] fix review comments.

alejandrounravel commented on airflow/models/taskinstance.py:
```diff
@@ -215,6 +215,49 @@ def try_number(self, value):
+            query = query.filter(TI.task_id == task_id)
+        if state:
+            query = query.filter(TI.state == state)
+        if state_not_equal:
```
Perhaps this should be "elif" so that "state" and "state_not_equal" don't clash with each other.

alejandrounravel reviewed (commented):
Suggesting tiny fix for an "elif". Other than that, LGTM, but need Airflow committers to approve.
